Support MCS (Queue-Based) Locks

babsingh opened the pull request:
TBD

babsingh commented:
Hi @gacholio and @fengxue-IS, I would like to request a review of the above design and implementation. I am experiencing hangs, situations where all threads wait and no thread is running. Please let me know if you notice issues in the design or implementation.

fyi - @DanHeidinga

fengxue-IS commented on thread/common/threadhelpers.cpp:
```diff
@@ -195,6 +195,246 @@ omrthread_spinlock_swapState(omrthread_monitor_t monitor, uintptr_t newState)
+
+lockReleased:
+	/* Pop the mcsNode from the thread's MCS node stack. */
+	self->mcsNodes->stackHead = mcsNode->stackNext;
```
If the lock acquire code (Line 281-288) is run right after `successor->blocked` is set to 0, and before this current line, does it break the following unlock operations after this one?

babsingh replied:
Every thread has its own `J9Thread->mcsNodes`, which contains the `stackHead` and `J9Pool` of MCS nodes. Each thread only reads/writes its own `J9Thread->mcsNodes`.

The thread, which will execute the lock acquire code after `successor->blocked` is set to 0, will access its own `J9Thread->mcsNodes`. No synchronization is required to access `J9Thread->mcsNodes`. Hence, the unlock operation should work fine.

gacholio commented:
`Assumption: The locks are acquired and released in order.`

This is certainly not a requirement at the moment - I know some of the VM exclusive code is very careful about lock release ordering (can't recall if it is always strictly ordered or not).

I think before we could make this kind of change, we'd have to enforce this with the current locking strategy and run with it for a considerable amount of time.

babsingh commented:
@gacholio In my debug branch, I have added code to handle out of order lock acquires/releases with MCS locks: https://github.com/babsingh/omr/commit/d111b93d4f90dd793285b598bb533e20ef6e7db8. Once I merge these changes in this pull request, I will void the assumption in the above design. Removing the assumption will make the design more universal and easily acceptable.

babsingh commented:
Updated the design above, and added latest changes to the pull request. Attn: @DanHeidinga.

babsingh commented:
**Update:**
- Pushed the latest changes.
- Addressed hangs. Reverted the changes to spin indefinitely in `omrthread_mcs_lock` until the lock is acquired. Now, threads block using a condition variable if spinning on a MCS lock exhausts.
- OMR compiles successfully, and all OMR testing passes especially the thread tests.
- OpenJ9 builds successfully with MCS locks. Now, it takes the same of amount of time to build OpenJ9 with MCS locks as the existing OpenJ9 builds. Before, it used to take 4-8 hours.
- Optimization for waking blocked threads is pending. Currently, all blocked threads are woken. Instead, we want to optimize by waking one or a reduced set of threads.
- While running OpenJ9 testing, hangs are noticed with MCS locks. Working on triaging/fixing these hangs.

babsingh commented:
**Update:**
- All OpenJ9 and OMR testing passes on Linux. No more hangs seen.

**TODO:**
- Support the new atomic exchange operation (`AtomicSupport::lockExchange`) on all platforms.
- Optimize the busy-wait loop for the MCS lock. Currently, the 3-tier busy-wait loop of TATAS lock is used. TATAS lock has a CAS operation in the 3-tier busy-wait loop but MCS doesn't have the CAS operation in the busy-wait loop. To compensate for the lack of CAS, the busy-wait loop will need to be modified for the MCS lock.
- Perform benchmarking.
- Implement concurrency restriction to handle the lock-waiter preemption issue. Only the next thread in the queue can acquire the lock. If this thread has to go through an expensive context switch, then the performance will be impacted. Concurrency restriction will reduce the size of the active thread set which will compete for the lock. Thus, reducing involuntary context switches.

babsingh commented:
**Cross platform implementation for the atomic exchange operation**
1. On GCC/GNU systems, the atomic exchange operation ([documentation](https://gcc.gnu.org/onlinedocs/gcc-4.1.0/gcc/Atomic-Builtins.html)) can be implemented using:
```
__sync_synchronize();
return __sync_lock_test_and_set(address, newValue);
```
2. On Windows VS (_MSC_VER), the atomic exchange operation is supported by [_InterlockedExchange](https://docs.microsoft.com/en-us/cpp/intrinsics/interlockedexchange-intrinsic-functions?view=vs-2019).
3. For XLC, the atomic exchange operation would be:
```
__fence();
__fetch_and_swap(address, newValue); //32-bit
OR
__fetch_and_swaplp(address, newValue); //64-bit
```

@fjeremic @zl-wang Is the above XLC implementation for the atomic exchange operation correct for AIX PPC and zOS?

@pshipton @rwy0717 Are the three implementations for the atomic exchange operation sufficient (GCC/GNU, _MSC_VER and XLC)? Will we need other implementations for covering all OpenJ9/OMR build environments?

0xdaryl commented:
> Are the three implementations for the atomic exchange operation sufficient (GCC/GNU, _MSC_VER and XLC)? Will we need other implementations for covering all OpenJ9/OMR build environments?

We build with Clang on macOS, so it is best to ensure you have an intrinsic that works there as well.

babsingh commented:
Created a standalone PR for the atomic exchange operation: https://github.com/eclipse/omr/pull/4086.

> We build with Clang on macOS, so it is best to ensure you have an intrinsic that works there as well.

I think the GNC implementation should work with Clang in general.

> Verify what instructions it generates to be 100% certain

Will verify in assembly.

zl-wang commented:
For performance reasons, different hardware hints are provided for atomic CompareAndSwap either for exclusive locking (i.e. with this CAS succeeding, nobody else can do anything meaningful on that memory slot until this thread does a write to release it)  or atomic update (i.e.  someone else can do atomic updates too after this CAS succeeds).  I assumed your use case is the latter for MCS. If yes, we have to check EH==0 hint is on the larx (load and reserve) instruction.  For exclusive, EH==1 (hardware will hold onto that cacheline longer).

zl-wang commented:
__fence() is ambiguous:  which kind of fence would you need?  and whether it is necessary?

xlC have __lwsync() --- light-weight sync;  and  __sync()  --- heavy-weight sync for your purpose.

I didn't see __fence() in the built-ins anyway.

All these are applicable to both AIX and Linux, if xlC is used.  They are xlC built-ins.

babsingh commented:
**Update:**

With https://github.com/eclipse/omr/pull/4596, the MCS (Mellor-Crummey Scott queue-based) STP (Spin-Then-Park) implementation is complete. I ran few benchmarks with Java workloads to compare the performance of the existing TATAS (Test-And-Test-And-Set) STP implementation with the new MCS STP implementation. Here are the results:
- Liberty Startup: ~5% regression with MCS STP.
- Liberty Footprint: Identical.
- [benchmarking-java-locks-with-counters](http://isuru-perera.blogspot.com/2016/05/benchmarking-java-locks-with-counters.html): Identical.
- VolanoMark: ~90% regression in throughput with MCS STP.

The Liberty startup and VolanoMark throughput regressions happen because of the lock-waiter preemption problem (spike in involuntary context switches). This can be solved by introducing unfairness to the MCS STP implementation via the [Malthusian](https://dl.acm.org/citation.cfm?id=3064203) variant. Currently working towards this goal.

fyi - @DanHeidinga
